### ISSUE TITLE
[Feature request] Instance Normalization

### ISSUE THREAD

**Issue by Zardinality** [tensorflow/tensorflow#11690]
[Instance Normalization](https://arxiv.org/abs/1607.08022) is recently widely used in style transfer and GAN, since it avoid the drawback of batch normalization which brings in-batch correlations. Tensorflow only has quantized version instance norm right now, but a full version is also easy to implement, I am interested in it. 
Also I wonder if it is better to implemented in C++ using eigen or simply add some lines in `nn_impls.py` with the benefit of being able to use `fused_batch_norm`. [Mxnet](https://github.com/dmlc/mxnet/blob/03b7d1402ab8d1142391c601fddbc4081632aa3c/src/operator/instance_norm-inl.h) does the first way while [Pytorch](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/instancenorm.py) use the latter method.

**Comment by StefanoD** [comment]
You can use Batch Renormalization in [Keras Contrib](https://github.com/farizrahman4u/keras-contrib).

**Comment by aselle** [comment]
@Zardinality, does batch renormalization satisfy your needs?

**Comment by Zardinality** [comment]
Batch renormalization is not exactly what I want, but I found instance normalization in [the same file](https://github.com/farizrahman4u/keras-contrib/commit/f0bb5becbb0585384890aefc51ffb11a4d832b2b). It is a pity that contrib part of keras isn't contained in tensorflow distribution.

**Comment by lllyasviel** [comment]
The implementation in https://github.com/farizrahman4u/keras-contrib is not exactly right.
According to the paper, the IN layer should perform differently in training/testing phase.
There needs a running_mean and a running_var.
The right Torch implementation for reference is here:
https://github.com/junyanz/CycleGAN/blob/master/util/InstanceNormalization.lua
**And it is the reason why Keras has no IN now.**

**Comment by the-moliver** [comment]
@lllyasviel @StefanoD  The instance normalization paper says: "The key idea is to replace batch normalization layers in the generator architecture with instance normalization layers, and to keep them at test time (as opposed to freeze and simplify them out as done for batch normalization)". I believe the running_mean and running_var in the lua implementations are being used to hack the batch normalization implementation to do instance normalization. Instance normalization is just contrast normalization and the keras-contrib implementation looks right to me. Please point out specific bugs if you identify them.

**Comment by lllyasviel** [comment]
I reviewed the code and finally found I was wrong. I am sorry for that. The keras implementation is good. 👍 
ps: maybe use default axis = -1 is better? Many keras users prefer directly writing "x = InstanceNormalization()(x);" when the last layer is a channel-last conv2d. In this case, it will take them many time to finally found out that they should add "axis=3".
I know that the default axis=None is prepared for dense layer. But the original keras prefers the -1 implementation, right?

**Comment by HectorAnadon** [comment]
I have a tensorflow model that I would like to add Instance Normalization. Is there a way to use [Keras Contrib](https://github.com/keras-team/keras-contrib) in tensorflow?
Thank you

**Comment by wandering007** [comment]
Are the implementations of Instance Normalization and Layer Normalization identical or similar?

**Comment by ValerioB88** [comment]
@lllyasviel interestingly enough, axis=None seems to provide better result of axis=-1 (on cycleGAN). I know this is any vague, but I thought it was worth it to put it out there.

**Comment by zhaoyingjun** [comment]
tf 2.0  replace the  api of Instance Normalization ?

**Comment by hfawaz** [comment]
So is Instance Normalization present in TF2.0 ?
